tasklist/client/src/Tasks/Task/Variables/VariableEditor/index.tsx

```diff
@@ -0,0 +1,314 @@
+ *
+ * If the Licensee is in breach of the Conditions, this Agreement, including the rights granted under it, will automatically terminate with immediate effect.
+ *
+ * SUBJECT AS SET OUT BELOW, THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ * NOTHING IN THIS AGREEMENT EXCLUDES OR RESTRICTS A PARTY’S LIABILITY FOR (A) DEATH OR PERSONAL INJURY CAUSED BY THAT PARTY’S NEGLIGENCE, (B) FRAUD, OR (C) ANY OTHER LIABILITY TO THE EXTENT THAT IT CANNOT BE LAWFULLY EXCLUDED OR RESTRICTED.
+ */
+
+import {RefObject} from 'react';
+import {
+  IconButton,
+  StructuredListBody,
+  StructuredListCell,
+  StructuredListHead,
+  StructuredListRow,
+  StructuredListWrapper,
+  TextInput,
+} from '@carbon/react';
+import {Popup, Close} from '@carbon/react/icons';
+import {Variable} from 'modules/types';
+import {Field, useFormState} from 'react-final-form';
+import {FieldArray} from 'react-final-form-arrays';
+import {DelayedErrorField} from '../../DelayedErrorField';
+import {LoadingTextarea} from '../LoadingTextarea';
+import {OnNewVariableAdded} from '../OnNewVariableAdded';
+import {
+  createVariableFieldName,
+  createNewVariableFieldName,
+} from '../createVariableFieldName';
+import {
+  validateValueJSON,
+  validateNameCharacters,
+  validateNameComplete,
+  validateDuplicateNames,
+  validateValueComplete,
+} from '../validators';
+import {mergeValidators} from '../validators/mergeValidators';
+import styles from './styles.module.scss';
+import cn from 'classnames';
+import {FormValues} from '../types';
+
+type Props = {
+  containerRef: RefObject<HTMLElement | null>;
+  variables: Variable[];
+  readOnly: boolean;
+  fetchFullVariable: (id: string) => void;
+  variablesLoadingFullValue: string[];
+  onEdit: (id: string) => void;
+};
+
+const CODE_EDITOR_BUTTON_TOOLTIP_LABEL = 'Open JSON code editor';
+
+function variableIndexToOrdinal(numberValue: number): string {
+  const realOrderIndex = (numberValue + 1).toString();
+
+  if (['11', '12', '13'].includes(realOrderIndex.slice(-2))) {
+    return `${realOrderIndex}th`;
+  }
+
+  switch (realOrderIndex.slice(-1)) {
+    case '1':
+      return `${realOrderIndex}st`;
+    case '2':
+      return `${realOrderIndex}nd`;
+    case '3':
+      return `${realOrderIndex}rd`;
+    default:
+      return `${realOrderIndex}th`;
+  }
+}
+
+const VariableEditor: React.FC<Props> = ({
```

Comment:
I don't think it's necessary to extract this into its own component

Reply:
I think this is a very large component and its hard to understand when it was embedded into the form code.

Reply:
The variable view is only used for debugging these days, it almost never change it anymore
I would just leave it as is

For the process view we can just wrap it around the tab component, no?